src/encode.ts

```diff
@@ -0,0 +1,477 @@
+				if (typeof newValue === "object") {
+					counters.refId--;
+				} else {
+					refs.delete(value);
+				}
+			} else {
+				{
+					let isIterable =
+						typeof (value as any)[Symbol.iterator] === "function";
+
+					if (isIterable) {
+						let isArray = Array.isArray(value);
+						let toEncode = isArray
+							? value
+							: Array.from(value as Iterable<unknown>);
+
+						encodeStack.push(
+							new EncodeFrame(2, "]", undefined) as EncodeFrameObj,
+						);
+						for (let i = (toEncode as unknown[]).length - 1; i >= 0; i--) {
+							encodeStack.push(
+								new EncodeFrame(
+									1,
+									i === 0 ? "" : ",",
+									(toEncode as unknown[])[i],
+								) as EncodeFrameObj,
+							);
+						}
+						chunks.push(
+							isArray
+								? "["
+								: value instanceof Set
+									? `${STR_SET}[`
+									: value instanceof Map
+										? `${STR_MAP}[`
+										: "[",
+						);
+						continue;
+					}
+				}
+
+				{
+					let pluginsLength = plugins.length;
+					for (let i = 0; i < pluginsLength; i++) {
+						let result = plugins[i](value);
+						if (Array.isArray(result)) {
+							encodeStack.push(
+								new EncodeFrame(1, STR_PLUGIN, result) as EncodeFrameObj,
+							);
+							continue encodeLoop;
+						}
+					}
+				}
+
+				encodeStack.push(new EncodeFrame(2, "}", undefined) as EncodeFrameObj);
+				{
+					let keys = Object.keys(value as object);
+					let end = keys.length;
+					let encodeFrames = new Array(end);
+					end -= 1;
+
+					for (let i = keys.length - 1; i >= 0; i--) {
+						let key = keys[i];
+						let prefix = i > 0 ? "," : "";
+						encodeFrames[end - i] = new EncodeFrame(
+							1,
+							`${prefix}${JSON.stringify(key)}:`,
+							(value as any)[key],
+						);
+					}
+
+					encodeStack.push(...encodeFrames);
+				}
+				chunks.push("{");
+			}
+		} else if (typeOfValue === "string") {
+			chunks.push(JSON.stringify(value));
+		} else if (typeOfValue === "number") {
+			if (nan(value)) {
+				chunks.push(STR_NaN);
+			} else if (value === POSITIVE_INFINITY) {
+				chunks.push(STR_INFINITY);
+			} else if (value === NEGATIVE_INFINITY) {
+				chunks.push(STR_NEGATIVE_INFINITY);
+			} else if (Object.is(value, -0)) {
+				chunks.push(STR_NEGATIVE_ZERO);
+			} else {
+				chunks.push(value.toString());
+			}
+		} else if (typeOfValue === "bigint") {
+			chunks.push(STR_BIGINT, value.toString());
+		} else if (typeOfValue === "symbol") {
+			let symbolKey = Symbol.keyFor(value as symbol);
+			if (typeof symbolKey === "string") {
+				chunks.push(STR_SYMBOL, JSON.stringify(symbolKey));
+			} else {
+				chunks.push(STR_UNDEFINED);
+			}
+		} else {
+			chunks.push(STR_UNDEFINED);
+		}
+	}
+}
+
+function prepareErrorForEncoding(error: Error, redactErrors: boolean) {
+	return {
+		name: redactErrors ? "Error" : error.name,
+		message: redactErrors ? STR_REDACTED : error.message,
+		stack: redactErrors ? STR_REDACTED : error.stack,
+		cause: error.cause,
```

Comment:
Shouldn't `cause` also be redacted? This might already happen through recursing the returned object, so in that case dismiss the mention :)
```suggestion
		cause: prepareErrorForEncoding(error.cause, redactErrors),
```

Reply:
Cause goes through the same encoding flow as the source error, leading to it being redacted if it's also an error instance.